File: 2015/09/solution.js
let distances = {}

const processInputLine = (line) => {
  let [, src, dest, dist] = line.match(/^(.*) to (.*) = (\d+)$/)
  dist = parseInt(dist)
  if (!distances[src]) distances[src] = {}
  if (!distances[dest]) distances[dest] = {}
  distances[src][dest] = dist
  distances[dest][src] = dist
}

const permuteDestinations = (destinations) => {
  if (destinations.length <= 1) return destinations
  let result = []
  for (let i = 0; i < destinations.length; i++) {
    const destination = destinations[i]
    destinations.splice(i, 1)
    permuteDestinations(destinations).forEach((permutation) => {
      result.push([ destination ].concat(permutation))
    })
    destinations.splice(i, 0, destination)
  }
  return result
}

const pathLength = (path) => {
  let length = 0
  for (let i = 1; i < path.length; i++) {
    length += distances[path[i - 1]][path[i]]
  }
  return length
}

const solve = () => {
  let minPathLength, maxPathLength
  permuteDestinations(Object.keys(distances)).forEach((path) => {
    const length = pathLength(path)
    if (!minPathLength || minPathLength > length) {
      minPathLength = length
    }
    if (!maxPathLength || maxPathLength < length) {
      maxPathLength = length
    }
  })
  console.log(`Part 1 (Min Path): ${minPathLength}`)
  console.log(`Part 2 (Max Path): ${maxPathLength}`)
}

require('readline').
    createInterface({ input: require('fs').createReadStream('input.txt') }).
    on('line', processInputLine).
    on('close', solve)

File: 2015/09/solution.rb
def permute_destinations(destinations)
  return [ destinations ] if destinations.size <= 1
  result = []
  destinations.each do |destination|
    destinations.delete(destination)
    permute_destinations(destinations).each do |permutation|
      result.push([ destination ] + permutation)
    end
    destinations.unshift destination
  end
  result
end

def path_length(distances, path)
  length = 0
  for i in 1...path.size
    length += distances[path[i - 1]][path[i]]
  end
  length
end

distances = {}

File.open('input.txt').each_line do |line|
  src, dest, dist = /^(.*) to (.*) = (\d+)$/.match(line)[1..3]
  dist = dist.to_i
  distances[src] = {} unless distances[src]
  distances[dest] = {} unless distances[dest]
  distances[src][dest] = dist
  distances[dest][src] = dist
end

max_path_length = 0
min_path_length = Float::INFINITY

permute_destinations(distances.keys).each do |path|
  length = path_length(distances, path)
  min_path_length = length if min_path_length > length
  max_path_length = length if max_path_length < length
end

puts 'Part 1', min_path_length
puts 'Part 2', max_path_length